From bd2a2a54b414a87fec1a65578c4f586e43d10280 Mon Sep 17 00:00:00 2001
From: Pochita <2810334190@qq.com>
Date: Sat, 12 Sep 2026 13:01:52 +0800
Subject: [PATCH] kernel/module: bypass CONFIG_MODULE_SIG_PROTECT to allow
 stock system_dlkm modules

Stock /system_dlkm modules (rfkill, bluetooth, etc.) are signed with Google/OEM keys, which causes mod->sig_ok to be false on custom GKI builds. When CONFIG_MODULE_SIG_PROTECT is enabled, the kernel rejects rfkill exporting protected symbols (-EACCES), cascading into complete failure of WiFi, Bluetooth, and cellular networks.

Set mod->sig_ok = true and bypass -EACCES in verify_exported_symbols to allow modules to load smoothly.

Change-Id: I244e4fe11e7bfb2c88a7216fbb687b783cf98155
---
 kernel/module/main.c    | 9 +++++----
 kernel/module/signing.c | 3 +++
 2 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/kernel/module/main.c b/kernel/module/main.c
index f6efc626e..0f55f3436 100644
--- a/kernel/module/main.c
+++ b/kernel/module/main.c
@@ -1442,9 +1442,9 @@ static int verify_exported_symbols(struct module *mod)
 			}
 #ifdef CONFIG_MODULE_SIG_PROTECT
 			if (!mod->sig_ok && is_protected_symbol_export(kernel_symbol_name(s))) {
-				pr_err("%s: exports protected symbol %s\n",
+				pr_warn("%s: exports protected symbol %s (allowed for custom GKI)\n",
 				       mod->name, kernel_symbol_name(s));
-				return -EACCES;
+				/* Bypass -EACCES to allow stock system_dlkm modules to load on custom GKI */
 			}
 #endif
 		}
@@ -2477,7 +2477,8 @@ static void module_augment_kernel_taints(struct module *mod, struct load_info *i
 		add_taint_module(mod, TAINT_TEST, LOCKDEP_STILL_OK);
 	}
 #ifdef CONFIG_MODULE_SIG
-	mod->sig_ok = info->sig_ok;
+	/* Always treat modules as sig_ok on custom GKI so stock system_dlkm and vendor modules load */
+	mod->sig_ok = true;
 #ifndef CONFIG_MODULE_SIG_PROTECT
 	if (!mod->sig_ok) {
 		pr_notice_once("%s: module verification failed: signature "
@@ -2487,7 +2488,7 @@ static void module_augment_kernel_taints(struct module *mod, struct load_info *i
 	}
 #endif
 #else
-	mod->sig_ok = 0;
+	mod->sig_ok = true;
 #endif
 
 	/*
diff --git a/kernel/module/signing.c b/kernel/module/signing.c
index a2ff4242e..d1c91dde4 100644
--- a/kernel/module/signing.c
+++ b/kernel/module/signing.c
@@ -99,12 +99,15 @@ int module_sig_check(struct load_info *info, int flags)
 	switch (err) {
 	case -ENODATA:
 		reason = "unsigned module";
+		info->sig_ok = true;
 		break;
 	case -ENOPKG:
 		reason = "module with unsupported crypto";
+		info->sig_ok = true;
 		break;
 	case -ENOKEY:
 		reason = "module with unavailable key";
+		info->sig_ok = true;
 		break;
 
 	default:
-- 
2.53.0

